File: recfun.hs
-- Assignment 1 in Haskell

pascalNaive c r | c == 0 || c == r = 1
pascalNaive c r                    = left + right
  where left  = pascalNaive (c-1) (r-1)
        right = pascalNaive   c   (r-1)

-- David verziojanak parafrazisa
pascalDavid c r = last $ iterate col start !! c
  where start        = take (r-c+1) $ repeat 1
        col [x]      = [x]
        col (x:y:ys) = x : (col $ x+y : ys)

-- Gyors verzio
pascal c r = rec r 1 1
  where rec _ k acc | k > c = acc
        rec n k acc         = rec (n-1) (k+1) (acc * n `div` k)

pascalRowString row = concat $ map ((++ " ") . show) xs
  where xs = [pascal col row | col <- [0,1..row]]

main = do
  putStrLn "Pascal's Triangle"
  mapM_ (putStrLn . pascalRowString) [0,1..10]

balance chars = rec chars 0
  where rec []         open = open == 0
        rec ('(':tail) open = rec tail (open+1)
        rec (')':tail) open = open > 0 && rec tail (open-1)
        rec (_:tail)   open = rec tail open

countChange 0    _              = 1
countChange _    []             = 0
countChange m c@(x:xs) | x <= m = countChange (m-x) c + countChange m xs
countChange m   (_:xs)          = countChange m xs

File: recfun.scala
package recfun
import common._

object Main {
  def main(args: Array[String]) {
    println("Pascal's Triangle")
    for (row <- 0 to 10) {
      for (col <- 0 to row)
        print(pascal(col, row) + " ")
      println()
    }
  }

  def pascal(c: Int, r: Int): Int =
    if (c == 0 || c == r) 1
    else pascal(c-1, r-1) + pascal(c, r-1)

  def balance(chars: List[Char]): Boolean = {
    def rec(chars: List[Char], open: Int): Boolean =
      if (chars.isEmpty) open == 0
      else if (chars.head == '(') rec(chars.tail, open + 1)
      else if (chars.head == ')') open > 0 && rec(chars.tail, open - 1)
      else rec(chars.tail, open)
    rec(chars, 0)
  }

  def countChange(money: Int, coins: List[Int]): Int =
    if (money == 0) 1
    else if (coins.isEmpty) 0
    else if (coins.head <= money)
      countChange(money - coins.head, coins) + countChange(money, coins.tail)
    else countChange(money, coins.tail)
}
